app/src/main/java/org/openobservatory/ooniprobe/activity/oonirun/OoniRunV2Activity.kt

```diff
@@ -0,0 +1,217 @@
+import org.openobservatory.ooniprobe.common.TestDescriptorManager
+import org.openobservatory.ooniprobe.common.ThirdPartyServices
+import org.openobservatory.ooniprobe.databinding.ActivityOoniRunV2Binding
+import org.openobservatory.ooniprobe.model.database.TestDescriptor
+import javax.inject.Inject
+
+/**
+ * Activity to handle a v2 link.
+ *
+ * A v2 link has the following format:
+ *
+ * 1. ooni://runv2/link_id
+ * 2. https://run.test.ooni.org/v2/link_id
+ *
+ * The activity is started when the user clicks on `Open Link in OONI Probe` or
+ * when the system recognizes this app can open this link and launches the app when a link is clicked.
+ *
+ * It fetches the descriptor from the link and starts the `AddDescriptorActivity`.
+ * If the link is invalid, it shows an error message and ends the activity.
+ *
+ * @see {@link org.openobservatory.ooniprobe.activity.OoniRunActivity} for v1 links.
+ */
+class OoniRunV2Activity : AbstractActivity() {
+    lateinit var binding: ActivityOoniRunV2Binding
+
+    @Inject
+    lateinit var descriptorManager: TestDescriptorManager
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        activityComponent.inject(this)
+        binding = ActivityOoniRunV2Binding.inflate(layoutInflater)
+        setContentView(binding.root)
+        onNewIntent(intent)
+    }
+
+    override fun onNewIntent(intent: Intent) {
+        super.onNewIntent(intent)
+        /**
+         * Check if we are starting the activity from a link [Intent.ACTION_VIEW].
+         * This is invoked when a v2 link is opened.
+         * @see {@link org.openobservatory.ooniprobe.activity.OoniRunActivity.newIntent}. for v1 links.
+         */
+        if (Intent.ACTION_VIEW == intent.action) {
+            manageIntent(intent)
+        } else {
+            // If the intent action is not `Intent.ACTION_VIEW`, end activity.
+            Toast.makeText(this, getString(R.string.Modal_Error), Toast.LENGTH_LONG).show()
+            finish()
+        }
+    }
+
+    /**
+     * Parses the intent data to extract the link.
+     * If the intent does not contain a link, show an error message and end the activity.
+     * If the intent contains a link, but it is not a supported link or has a non-numerical `link_id`,
+     * show an error message and end the activity.
+     * If the intent contains a link, but the `link_id` is zero,
+     * show an error message and end the activity.
+     * If the intent contains a link with a valid `link_id`,
+     * fetch the descriptor from the link and start the `AddDescriptorActivity`.
+     *
+     * @param intent The intent data.
+     */
+    private fun manageIntent(intent: Intent) {
+        // If the intent does not contain a link, do nothing.
+        val uri = intent.data ?: finishWithError().run { return }
+        // If the §intent contains a link, but it is not a supported link or has a non-numerical `link_id`.
```

Comment:
```suggestion
        // If the intent contains a link, but it is not a supported link or has a non-numerical `link_id`.
```

is the paragraph symbol intended?